libsolidity/codegen/ContractCompiler.h

```diff
@@ -123,19 +125,35 @@ class ContractCompiler: private ASTConstVisitor
 	/// @returns the runtime assembly for clone contracts.
 	eth::AssemblyPointer cloneRuntime() const;
 
+	/// Frees the variables of a certain scope (to be used when leaving).
+	void popScopedVariables(ASTNode const* _node);
+
+	/// Pops _amount slots from the stack and jumps to _jumpTo.
+	/// Also readjusts the stack offset to the original value.
+	void popAndJump(unsigned _amount, eth::AssemblyItem const& _jumpTo);
+
+	/// Sets the stack height for the visited loop.
+	void visitLoop(BreakableStatement const* _loop);
+
 	bool const m_optimise;
 	/// Pointer to the runtime compiler in case this is a creation compiler.
 	ContractCompiler* m_runtimeCompiler = nullptr;
 	CompilerContext& m_context;
-	std::vector<eth::AssemblyItem> m_breakTags; ///< tag to jump to for a "break" statement
-	std::vector<eth::AssemblyItem> m_continueTags; ///< tag to jump to for a "continue" statement
-	/// Tag to jump to for a "return" statement, needs to be stacked because of modifiers.
-	std::vector<eth::AssemblyItem> m_returnTags;
+	/// Tag to jump to for a "break" statement and the stack height after freeing the local loop variables.
+	std::vector<std::pair<eth::AssemblyItem, unsigned>> m_breakTags;
+	/// Tag to jump to for a "continue" statement and the stack height after freeing the local loop variables.
+	std::vector<std::pair<eth::AssemblyItem, unsigned>> m_continueTags;
+	/// Tag to jump to for a "return" statement and the stack height after freeing the local function or modifier variables.
+	/// Needs to be stacked because of modifiers.
```

Comment:
I think in a future refactor, we should collect all these fields that depend on the modifier depth and put them in a single struct or even instantiate a new compiler class.

Reply:
Makes sense, I'll create an issue to remember that.